README.md
[![Tsacdop Banner][]][google play]

[![github action][]][github action link]
[![GitHub Release][]][github release - recent]
[![Github Downloads][]][github release - recent]
[![Localizely][]][localizely - website]
[![style: effective dart][]][effective dart pub]
[![License badge][]][license]

## About

Enjoy podcasts with Tsacdop.

Tsacdop is a podcast player developed with Flutter, a clean, simply beautiful and friendly app, which is also free and open source.

Credit to Flutter team and all involved plugins, especially [webfeed](https://github.com/witochandra/webfeed) , [Just_Audio](https://pub.dev/packages/just_audio) and [Provider](https://pub.dev/packages/provider).

The podcast search engine is powered by [ListenNotes](https://listennotes.com) & [PodcastIndex](https://podcastindex.org/).

## Features

* Podcast group management
* Playlists support
* Sleep timer / speed setting
* OPML file export and import
* Auto syncing in background
* Listening and subscription history record
* Dark mode / accent color
* Download for offline play
* Auto download new episodes / auto delete outdated downloads
* Settings backup
* Skip silence
* Boost volume

More to come...

## Preview

| Home Page | Group | Podcast | Episode| Dark Mode |
| ----- | ----- | ----- | ------ | ----- |
|![][Homepage ScreenShot]|![][Group Screenshot] | ![][Podcast Screenshot] | ![][Episode Screenshot]| ![][Darkmode Screenshot] |

## Localization

Please [Email](mailto:<tsacdop.app@gmail.com>) me you'd like to contribute to support more languages!

Credit to [Localizely](https://localizely.com/) for kind support to open source projects.

### ![English]

### ![Chinese Simplified]

### ![French] 

### ![Spanish]

### ![Portuguese]

## License

Tsacdop is licensed under the [GPL v3.0](https://github.com/stonega/tsacdop/blob/master/LICENSE) license.

## Build

1. If you don't have Flutter SDK installed, please visit offcial [Flutter][Flutter Install] site.
2. Fetch latest sorce code from master branch.

``` 
git clone https://github.com/stonega/tsacdop.git
```

3. Add api search api configure file.  

Tsacdop uses ListenNotes API 1.0 pro to search for podcasts, which is not free, so I can not expose the API key in the repo.
If you want to build the app, you need to create a new file named `.env.dart` in lib folder. Add the following code to `.env.dart`. If you don't have ListenNotes api key, just keep apiKey ''. Than the app will only support the PodcastIndex search.
You can get your own ListenNotes API key on [ListenNotes](https://www.listennotes.com/api/), remember that you need to get pro plan API, because basic plan dosen't provide rss link for serach result. 

``` dart
final environment = {"apiKey":""};
```

4. Run the app with Android Studio or Visual Studio. Or the command line.

``` 
flutter pub get
flutter run
```

## Contribute 

If you have an issue or found a bug, please raise a GitHub issue. Pull requests are also welcome.

## Archetecture

### Plugins

* Local storage
  + sqflite
  + shared_preferences
* Audio
  + just_audio
  + audio_service
* State management
  + provider
* Download
  + flutter_downloader
* Background task
  + workmanager

### Directory Structure

``` 
UI
src
└──home
   ├──home.dart [Homepage]
   ├──searc_podcast.dart [Search Page]
   └──playlist.dart [Playlist Page]
└──podcasts
   ├──podcast_manage.dart [Group Page]
   └──podcast_detail.dart [Podcast Page]
└──episodes
   └──episode_detail.dart [Episode Page]
└──settings
   └──setting.dart [Setting Page]

STATE
src
└──state
   ├──audio_state.dart [Audio State]
   ├──download_state.dart [Episode Download]
   ├──podcast_group.dart [Podcast Groups]
   ├──refresh_podcast.dart [Episode Refresh]
   └──setting_state.dart [Setting]

Service
src
└──service
   ├──api_service.dart [Podcast Search]
   ├──gpodder_api.dart [Gpodder intergate]
   └──ompl_builde.dart [OMPL export]
```

## Known Issue

* Playlist is unstable

## Contact

You can reach out to me directly at [tsacdop.app@gmail.com](mailto:<tsacdop.app@gmail.com>).

Or you can join our [Telegram Group](https://t.me/joinchat/Bk3LkRpTHy40QYC78PK7Qg).

## Getting Started with Flutter

This project is a starting point for a Flutter application.

Here are a few resources to get you started if this is your first Flutter project:

* [Lab: Write your first Flutter app](https://flutter.dev/docs/get-started/codelab)
* [Cookbook: Useful Flutter samples](https://flutter.dev/docs/cookbook)

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials, samples, guidance on mobile development, and a full API reference.

[Flutter Install]: https://flutter.dev/docs/get-started/install
[tsacdop banner]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/banner.png
[build status - cirrus]: https://circleci.com/gh/stonega/tsacdop/tree/master.svg?style=shield
[github action]: https://github.com/stonega/tsacdop/workflows/Flutter%20Build/badge.svg
[github action link]: https://github.com/stonega/tsacdop/actions
[build status ]: https://circleci.com/gh/stonega/tsacdop/tree/master
[github release]: https://img.shields.io/github/v/release/stonega/tsacdop
[github release - recent]: https://github.com/stonega/tsacdop/releases
[github downloads]: https://img.shields.io/github/downloads/stonega/tsacdop/total?color=%230000d&label=downloads
[localizely]: https://img.shields.io/badge/dynamic/json?color=%2326c6da&label=localizely&query=%24.languages.length&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus
[English]: https://img.shields.io/badge/dynamic/json?style=for-the-badge&color=%2323CCC6&label=English&query=%24.languages%5B3%5D.reviewedProgress&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus&suffix=%
[Chinese Simplified]: https://img.shields.io/badge/dynamic/json?style=for-the-badge&color=%2323CCC6&label=Chinese%20Simplified&query=%24.languages%5B2%5D.reviewedProgress&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus&suffix=%
[French]: https://img.shields.io/badge/dynamic/json?style=for-the-badge&color=%2323CCC6&label=French(ppp)&query=%24.languages%5B5%5D.reviewedProgress&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus&suffix=%
[Spanish]: https://img.shields.io/badge/dynamic/json?style=for-the-badge&color=%2323CCC6&label=Spanish(Joel)&query=%24.languages%5B7%5D.reviewedProgress&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus&suffix=%
[Portuguese]: https://img.shields.io/badge/dynamic/json?style=for-the-badge&color=%2323CCC6&label=portuguese(Bruno)&query=%24.languages%5B9%5D.reviewedProgress&url=https%3A%2F%2Fapi.localizely.com%2Fv1%2Fprojects%2Fbde4e9bd-4cb2-449b-9de2-18f231ddb47d%2Fstatus&suffix=%
[localizely - website]: https://localizely.com/
[google play - icon]: https://img.shields.io/badge/google-playStore-%2323CCC6
[google play]: https://play.google.com/store/apps/details?id=com.stonegate.tsacdop
[Homepage ScreenShot]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/1585893838840.png
[Group Screenshot]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/1585894051734.png
[Podcast Screenshot]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/1585893877702.png
[Episode Screenshot]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/1585896237809.png
[Darkmode Screenshot]: https://raw.githubusercontent.com/stonega/tsacdop/master/preview/1585893920721.png
[style: effective dart]: https://img.shields.io/badge/style-effective_dart-40c4ff.svg
[effective dart pub]: https://pub.dev/packages/effective_dart
[license]: https://github.com/stonega/tsacdop/blob/master/LICENSE
[License badge]: https://img.shields.io/badge/license-GPLv3-yellow.svg


CHANGELOG.md
# Tsacdop Changelog
## 0.6.0
Release date 2021/2/27

### New features
* Load local folder as playlist
* Option to always hide new mark for podcast.
* Memory use reduced a lot
* Add reddit link

## 0.5.9

Release date 2021/2/11
### Bug fix

* Some podcasts subscribe failed.
* Update fastlane screenshots.

## 0.5.8
Release date 2021/2/9

### New features

* Artwork shownotes and chapters in player panel.
* Play without subscription, inside search result directly.

### Bug fix

* Episode download failed with '/' in title.

## 0.5.7
Release date 2021/1.31
For frdoid

## 0.5.6
Release date 2021/1.28

### Bug fix
* Image saved wrongly after subscribe new podcast.

## 0.5.5
Release date 2021/1/25

### Bugs fix
* Position not saved after player stpoped.
* Wrongly remove episodes from playlist when get back from playlist page.
* Player crash on OP5T.

### Minor change
* Playlist page UI update.

## 0.5.4
Release date 2021/1/9

### New feature
* Support podcastindex namespace podcast:funding, 

### Minor change
* Podcast detail page UI update. 

### Bug fix
* Custom rss link subscribe error.

## 0.5.3
Release date 2021/1/7

### Minor change
* Option to open playlist page by default.
* Option to open podcast list page instead of group page.
* Add remove new mark buttong in podcast page.

## 0.5.2

Release date 2021/1/3
First release in 2021

### New features
* Custom playlist support.
* Playlist page redesign.
* Podcast detail page and group management page UI update.
* Add remove new mark button in recent tab antion bar.
* Support time setting in sleep mode.

### Minor change
* Remove add new to playlist button.
* Groups management perfomance improved.
* Podcast page opennning animation improved.

## 0.5.1

Release date 2020/11/1

### Bug fixed

* ListenNotes search error.
* Sync podcasts marked as never auto update.
* Download failed when network is busy.

### Minor change

* User can add episode to the top of playlist in popup menu.
* Show podcast name in episode detail page.
* Tap pocdast avatar in episode card to open podcast page.

## 0.5.0

Release date 2020/10/13

### New fewtures

* Support multi select on recent and favorite tab.
* Select all/ select before/ select after.
* Option to delete episode download file when played.
* OPtion to mark as listened after skipped.

### Bug fixed

* Feed pubdate parse error.
* Episodes load with initial position failed.

### Minor changes

* Single colume layout update.
* About page UI update.
* More smooth animation when open podcast detail page.
* Change sort by button style in podcast detail page.
* Auto rewind 3 seconds when resuming from paused state.

## 0.4.20

Release date 2020/10/3

### Bug fixed

* Rss feed parse error.

## 0.4.19

Release date 2020/10/1

### New features

* Set podcastindex as  default search engine.
* Option to hide podcast discovery in search page.
* Italian translation support, thanks Edoardo.

### Bug fixed

* Mark all listened error.

## 0.4.18

Release date 2020/9/27

### New features

* Support gpodder.net sync.
* Portuguese translation, thanks Bruno.
* Turn off auto update for podcast.
* Pull to refresh in recent tab, supports group update.

### Minor changes

* Longpress 'see all' to open full podcast list.

## 0.4.17

Release date 2020/9/16

### Bug fixed

* Remove notification after app removed from recent.

## 0.4.16

Release date 2020/9/15

### New features

* Discovery feature in search page.
* Multi select in podcast page.
* Customize the speed options available.

### Bugs fixed

* Fix download error when podcast name includes /.
* Make the group name editable directly.
* Fixed shownote timestamp click error.

### Minor changes

* Update donate button UI.

## 0.4.15

Release date 2020/8/30

### New features

* Option to change notification panel layout.
* Option to change show notes font style.
* Option to hide listened default.
* Change skip next/previous to fastForward/rewind on headset click.

### Bugs fixed

* Download error when filename too long.

### Minor change

* Update download button style and downloaded indicator style.
* Add 1.1 to speed setting.
* Add 5s to skip seconds setting.

## 0.4.14

Release date 2020/8/20
Only for izzyonandroid.

## 0.4.13

Release date 2020/8/19

### Bugs fixed

* Downloaded episode play error, you might need to redownload the episode.

## 0.4.12

Release date 2020/8/15

### Bugs fixed

* Crash when reorder episodes or podcasts.
* Popup menu setting import bug.
* Default language failed to load.

### Minor changes

* Change language|feedback|podcast settings to button sheet.
* Add history in home playlist button.
* History page UI improved.

## 0.4.11

Release date 2020/8/12

### New features

* Boost volume. You can change boost level in settings.
* You can tap time stamp to skip instantly in shownote when the episode is playing.
* Add history list in playlist page.
* You can also mark not listened now.

### Minor change

* Improved time picker UI.
* Add episode setting page.

### Bugs fix

* Play record didn't saved after stop playing.
* Network error message didn't disapear after skip to next.
* Fireside avatar load error.

## 0.4.10

Release date 2020/8/6

### Bugs fixed

* Episdoe date parse error.
* Play from start after interrupt.
* Playlist in player unstable.
* Language setting not saved after restart app.

### Minor change

* Fast forward and rewind buttons UI changed.

## v0.4.9

Release date 2020/8/1

### New features

* Player UI redesign.
* Added player height setting.
* Added skip silence feature.

### Bugs fixed

* Language fixs. (Thanks to Atrate)
* Make app movable to SD card. (Thanks to Atrate)

### Minor change

* Episode page open animation improved.

lib/home/about.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:line_icons/line_icons.dart';

import '../util/extension_helper.dart';
import '../widgets/custom_widget.dart';

const String version = '0.6.0';

class AboutApp extends StatefulWidget {
  @override
  _AboutAppState createState() => _AboutAppState();
}

class _AboutAppState extends State<AboutApp> {
  ScrollController _scrollController;
  bool _scroll;
  @override
  void initState() {
    super.initState();
    _scroll = false;
    _scrollController = ScrollController()
      ..addListener(() {
        if (_scrollController.offset > 0 && !_scroll && mounted) {
          setState(() => _scroll = true);
        }
        if (_scrollController.offset <= 0 && _scroll && mounted) {
          setState(() => _scroll = false);
        }
      });
  }

  Widget _listItem(
          BuildContext context, String text, IconData icons, String url) =>
      InkWell(
        onTap: () => url.launchUrl,
        child: Container(
          height: 50.0,
          padding: EdgeInsets.symmetric(horizontal: 20.0),
          alignment: Alignment.centerLeft,
          decoration: BoxDecoration(
            border: Border(
              bottom: Divider.createBorderSide(context),
            ),
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              Icon(icons, color: Theme.of(context).accentColor),
              Padding(
                padding: EdgeInsets.symmetric(horizontal: 10),
              ),
              Text(text),
            ],
          ),
        ),
      );

  Widget _translatorInfo(BuildContext context, {String name, String flag}) =>
      Container(
        height: 50.0,
        padding: EdgeInsets.symmetric(horizontal: 20.0),
        alignment: Alignment.centerLeft,
        decoration: BoxDecoration(
          border: Border(
            bottom: Divider.createBorderSide(context),
          ),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Icon(LineIcons.user, color: Theme.of(context).accentColor),
            Padding(
              padding: EdgeInsets.symmetric(horizontal: 10),
            ),
            Expanded(
                child: Text(
              name,
              maxLines: 1,
              overflow: TextOverflow.fade,
            )),
            if (flag != null)
              ClipRRect(
                borderRadius: BorderRadius.circular(4),
                child: Image(
                  image: AssetImage('assets/$flag.png'),
                  height: 20,
                  width: 30,
                  fit: BoxFit.cover,
                ),
              ),
          ],
        ),
      );

  @override
  Widget build(BuildContext context) {
    OverlayEntry _createOverlayEntry(TapDownDetails detail) {
      // RenderBox renderBox = context.findRenderObject();
      var offset = detail.globalPosition;
      return OverlayEntry(
        builder: (constext) => Positioned(
          left: offset.dx - 5,
          top: offset.dy - 120,
          child: Container(
              width: 20,
              height: 120,
              color: Colors.transparent,
              alignment: Alignment.topCenter,
              child: HeartSet(height: 120, width: 20)),
        ),
      );
    }

    final s = context.s;
    return AnnotatedRegion<SystemUiOverlayStyle>(
      value: SystemUiOverlayStyle(
        statusBarIconBrightness: Theme.of(context).accentColorBrightness,
        systemNavigationBarColor: Theme.of(context).primaryColor,
        systemNavigationBarIconBrightness:
            Theme.of(context).accentColorBrightness,
      ),
      child: Scaffold(
        backgroundColor: Theme.of(context).primaryColor,
        appBar: AppBar(
          title: Text(s.homeToprightMenuAbout),
          leading: CustomBackButton(),
          elevation: _scroll ? 1 : 0,
        ),
        body: SafeArea(
          child: ScrollConfiguration(
            behavior: NoGrowBehavior(),
            child: SingleChildScrollView(
              controller: _scrollController,
              scrollDirection: Axis.vertical,
              child: Padding(
                padding: const EdgeInsets.all(20.0),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.start,
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Container(
                      height: 110.0,
                      alignment: Alignment.center,
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        mainAxisSize: MainAxisSize.min,
                        children: <Widget>[
                          Image(
                            image: AssetImage('assets/logo.png'),
                            height: 80,
                          ),
                          Text(s.version(version)),
                        ],
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.all(20),
                      child: Text(
                        'Tsacdop is a podcast player built with flutter, a clean, simply beautiful and friendly app.',
                        textAlign: TextAlign.center,
                      ),
                    ),
                    Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      crossAxisAlignment: CrossAxisAlignment.center,
                      children: [
                        TextButton(
                          onPressed: () =>
                              'https://tsacdop.stonegate.me/#/privacy'
                                  .launchUrl,
                          style: TextButton.styleFrom(
                              primary: context.accentColor,
                              textStyle:
                                  TextStyle(fontWeight: FontWeight.bold)),
                          child: Text(
                            s.privacyPolicy,
                          ),
                        ),
                        Container(
                          margin: const EdgeInsets.symmetric(horizontal: 5),
                          height: 4,
                          width: 4,
                          decoration: BoxDecoration(
                              color: context.accentColor,
                              shape: BoxShape.circle),
                        ),
                        TextButton(
                          onPressed: () =>
                              'https://tsacdop.stonegate.me/#/changelog'
                                  .launchUrl,
                          style: TextButton.styleFrom(
                              primary: context.accentColor,
                              textStyle:
                                  TextStyle(fontWeight: FontWeight.bold)),
                          child: Text(s.changelog,
                              style: TextStyle(color: context.accentColor)),
                        ),
                      ],
                    ),
                    Padding(
                      padding: EdgeInsets.only(top: 20.0, bottom: 10.0),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.start,
                        mainAxisSize: MainAxisSize.min,
                        children: <Widget>[
                          _listItem(context, 'Twitter @tsacdop',
                              LineIcons.twitter, 'https://twitter.com/tsacdop'),
                          _listItem(context, 'GitHub', LineIcons.alternateGithub,
                              'https://github.com/stonega/tsacdop'),
                          _listItem(context, 'Telegram', LineIcons.telegram,
                              'https://t.me/joinchat/Bk3LkRpTHy40QYC78PK7Qg'),
                          _listItem(context, 'Reddit', LineIcons.redditLogo,
                              'https://www.reddit.com/r/Tsacdop'),
                          Center(
                            child: SizedBox(
                              width: 200,
                              child: ElevatedButton(
                                onPressed: () =>
                                    'https://www.buymeacoffee.com/stonegate'
                                        .launchUrl,
                                style: ElevatedButton.styleFrom(
                                    primary: context.accentColor),
                                child: Container(
                                  height: 30.0,
                                  padding:
                                      EdgeInsets.symmetric(horizontal: 4.0),
                                  alignment: Alignment.center,
                                  child: Row(
                                    mainAxisAlignment: MainAxisAlignment.center,
                                    mainAxisSize: MainAxisSize.min,
                                    children: <Widget>[
                                      Text('Buy Me A Coffee',
                                          style: TextStyle(
                                              fontWeight: FontWeight.bold)),
                                      SizedBox(width: 10),
                                      Image(
                                        image: AssetImage(
                                            'assets/buymeacoffee.png'),
                                        height: 20,
                                        fit: BoxFit.fitHeight,
                                      ),
                                    ],
                                  ),
                                ),
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                    Padding(
                      padding: EdgeInsets.all(10.0),
                    ),
                    Padding(
                      padding: EdgeInsets.only(top: 20.0, bottom: 10.0),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.start,
                        mainAxisSize: MainAxisSize.min,
                        children: <Widget>[
                          Row(
                            children: [
                              SizedBox(width: 25),
                              Text(
                                s.translators,
                                style: TextStyle(
                                    color: Theme.of(context).accentColor,
                                    fontWeight: FontWeight.bold),
                              ),
                              SizedBox(width: 2),
                              Icon(Icons.favorite, color: Colors.red, size: 20),
                            ],
                          ),
                          _translatorInfo(context, name: 'Atrate'),
                          _translatorInfo(context, name: 'ppp', flag: 'fr'),
                          _translatorInfo(context,
                              name: 'Joel Israel', flag: 'mx'),
                          _translatorInfo(context,
                              name: 'Bruno Pinheiro', flag: 'pt'),
                          _translatorInfo(context,
                              name: 'Edoardo Maria Elidoro', flag: 'it'),
                           _translatorInfo(context,
                              name: 'Murat T. Akyuz', flag: 'tr'),
                        ],
                      ),
                    ),
                    //Spacer(),
                    Padding(
                      padding: EdgeInsets.symmetric(vertical: 10),
                    ),
                    Container(
                      height: 50,
                      alignment: Alignment.center,
                      child: GestureDetector(
                        onTapDown: (detail) async {
                          OverlayEntry _overlayEntry;
                          _overlayEntry = _createOverlayEntry(detail);
                          Overlay.of(context).insert(_overlayEntry);
                          await Future.delayed(Duration(seconds: 2));
                          _overlayEntry?.remove();
                        },
                        child: Row(
                          crossAxisAlignment: CrossAxisAlignment.center,
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: <Widget>[
                            Image.asset(
                              'assets/text.png',
                              height: 25,
                            ),
                            Padding(
                              padding: EdgeInsets.symmetric(horizontal: 5),
                            ),
                            Icon(
                              Icons.favorite,
                              color: Colors.blue,
                            ),
                            Padding(
                              padding: EdgeInsets.symmetric(horizontal: 5),
                            ),
                            FlutterLogo(
                              size: 18,
                            ),
                          ],
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
